File: analysis/hwk2-analysis.R

# Meta --------------------------------------------------------------------

## Title:         Econ/HLTH 470 Homework 2 Answers
## Author:        Ian McCarthy
## Date Created:  2/20/2020
## Date Edited:   2/8/2023
## Description:   This file renders/runs all relevant R code for the assignment


# Preliminaries -----------------------------------------------------------

if (!require("pacman")) install.packages("pacman")
pacman::p_load(tidyverse, ggplot2, dplyr, lubridate, stringr, readxl, data.table, gdata, scales,
               ggbeeswarm, MatchIt)


# Read data and set workspace for knitr -------------------------------
hcris.data.v1996 <- readRDS('data/HCRIS_Data_v1996.rds')
hcris.data.v2010 <- readRDS('data/HCRIS_Data_v2010.rds')
hcris.data <- readRDS('data/HCRIS_Data.rds')

# Create objects for markdown ---------------------------------------------

version.1996 <- hcris.data.v1996 %>% group_by(year) %>% summarize(count_1996=n())
version.2010 <- hcris.data.v2010 %>% group_by(year) %>% summarize(count_2010=n())
version.dat <- version.1996 %>%
  full_join(version.2010, by='year')

final.hcris.v1996 = hcris.data.v1996 %>%
  mutate(hvbp_payment=NA, hrrp_payment=NA)

## combine v1996 and v2010 hcris forms, and sort by provider_number/year
final.hcris=rbind(final.hcris.v1996,hcris.data.v2010) %>%
  mutate(fy_end=mdy(fy_end),fy_start=mdy(fy_start),
         date_processed=mdy(date_processed),date_created=mdy(date_created),
         tot_discounts=abs(tot_discounts), hrrp_payment=abs(hrrp_payment)) %>%
  mutate(fyear=year(fy_end)) %>%
  arrange(provider_number,fyear) %>%
  select(-year)

## count of hospitals/provider_number by year
dup.count <- final.hcris %>% 
  add_count(provider_number, fyear, name="total_reports") %>%
  mutate(dup_report = (total_reports>1)) %>%
  group_by(fyear) %>%
  summarize(duplicates=sum(dup_report))

fig.dup <- dup.count %>%
  ggplot(aes(x=as.factor(fyear), y=duplicates, group=1)) + geom_line() +
  labs(
    x="Year",
    y="Number of Hospitals",
    title=""
  ) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 70, hjust=1))  

fig.unique <- hcris.data %>% group_by(year) %>%
  summarize(hosp_count=n()) %>%
  ggplot(aes(x=as.factor(year), y=hosp_count, group=1)) + 
  geom_line() +
  labs(
    x="Year",
    y="Number of Hospitals",
    title=""
  ) + theme_bw() +
  scale_y_continuous(labels = comma) +
  theme(axis.text.x = element_text(angle = 70, hjust=1))

## violin plot of total charges
charge.data <- hcris.data %>%
  group_by(year) %>% 
  mutate(tot_charges_low=quantile(tot_charges, probs=0.01, na.rm=TRUE),
            tot_charges_high=quantile(tot_charges, probs=0.99, na.rm=TRUE)) %>%
  filter(tot_charges<tot_charges_high, tot_charges>tot_charges_low,
         !is.na(tot_charges), year>1997) %>%
  mutate(log_charge=log(tot_charges))

fig.totcharge <- charge.data %>%
  ggplot(aes(x=as.factor(year), y=log_charge)) +
  geom_quasirandom(alpha=0.05, width=0.2, color = "#eba487") +
  stat_summary(fun='median', geom='point', size=2, color = "#abdcf1") +
  stat_summary(fun='median', geom='line', aes(group=1), size=1.1, color = "#abdcf1") +
  labs(
    x="Year",
    y="Log $",
    title=""
  ) + 
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 70, hjust=1))



## Violin plot of prices
price.data <- hcris.data %>%
  mutate( discount_factor = 1-tot_discounts/tot_charges,
          price_num = (ip_charges + icu_charges + ancillary_charges)*discount_factor - tot_mcare_payment,
          price_denom = tot_discharges - mcare_discharges,
          price = price_num/price_denom) %>%
  filter(price_denom>100, !is.na(price_denom), 
         price_num>0, !is.na(price_num),
         price<100000, 
         beds>30, !is.na(beds))

fig.prices <- price.data %>%
  ggplot(aes(x=as.factor(year), y=price)) +
  geom_quasirandom(alpha=0.05, width=0.2, color = "#eba487") +
  stat_summary(fun='median', geom='point', size=2, color = "#abdcf1") +
  stat_summary(fun='median', geom='line', aes(group=1), size=1.1, color = "#abdcf1") +
  labs(
    x="Year",
    y="Dollars",
    title=""
  ) + 
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 70, hjust=1))

## Penalty graph
pen.data <- price.data %>%
  mutate( hvbp_payment = ifelse(is.na(hvbp_payment),0,hvbp_payment),
          hrrp_payment = ifelse(is.na(hrrp_payment),0,abs(hrrp_payment)),
          hrrp_penalty = as.numeric((hrrp_payment>0)),
          penalty = as.numeric((hvbp_payment-hrrp_payment<0)))

fig.pen <- pen.data %>% group_by(year) %>%
  ggplot(aes(x=as.factor(year), y=hrrp_penalty, group=1)) +
  stat_summary(fun="mean", geom="line", na.rm=TRUE) +
  labs(
    x="Year",
    y="Percent Penalized",
    title=""
  ) + 
  scale_y_continuous(labels=percent, limits=c(0,1)) +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 70, hjust=1))



## Final data for analysis
pen.data.2012 <- pen.data %>% filter(year==2012) %>% ungroup() %>%
  mutate(beds_q1 = quantile(beds, probs=0.25, na.rm=TRUE),
         beds_q2 = quantile(beds, probs=0.50, na.rm=TRUE),
         beds_q3 = quantile(beds, probs=0.75, na.rm=TRUE),
         beds_q4 = max(beds, na.rm=TRUE)) %>%
  mutate(bed_size1 = ifelse(beds<beds_q1,1,0),
         bed_size2 = ifelse(beds>= beds_q1 & beds<beds_q2,1,0),
         bed_size3 = ifelse(beds>= beds_q2 & beds<beds_q3,1,0),
         bed_size4 = ifelse(beds>  beds_q3 & beds<beds_q4,1,0),
         bed_quart = 1*bed_size1 + 2*bed_size2 + 3*bed_size3 + 4*bed_size4) %>%
  filter(bed_quart>0)


## average price by penalty status
avg.pen1 <- pen.data.2012 %>%
  group_by(penalty) %>%
  summarize(mean_price=mean(price))

avg.pen <- pen.data.2012 %>%
  group_by(penalty, bed_quart) %>% 
  summarize(mean_price=mean(price))

avg.pen.tab <- pivot_wider(avg.pen, names_from=penalty, values_from=mean_price, names_prefix="price_")


## matching
match.inv <- Matching::Match(Y=pen.data.2012$price,
                Tr=pen.data.2012$penalty,
                X= (pen.data.2012 %>% select(bed_size1, bed_size2, bed_size3)),
                M=1,
                Weight=1,
                estimand="ATE")

match.mah <- Matching::Match(Y=pen.data.2012$price,
                          Tr=pen.data.2012$penalty,
                          X= (pen.data.2012 %>% select(bed_size1, bed_size2, bed_size3)),
                          M=1,
                          Weight=2,
                          estimand="ATE")

  
## Propensity scores and IPW
logit.model <- glm(penalty ~ bed_size1 + bed_size2 + bed_size3, 
                   family=binomial, 
                   data=pen.data.2012)
ps <- fitted(logit.model)

pen.data.2012 <- pen.data.2012 %>%
  mutate(ipw = case_when(
    penalty==1 ~ 1/ps,
    penalty==0 ~ 1/(1-ps),
    TRUE ~ NA_real_
  ))

mean.t1 <- pen.data.2012 %>% filter(penalty==1) %>%
  select(price, ipw) %>% summarize(mean_p=weighted.mean(price,w=ipw))
mean.t0 <- pen.data.2012 %>% filter(penalty==0) %>%
  select(price, ipw) %>% summarize(mean_p=weighted.mean(price,w=ipw))
ipw.diff <- mean.t1$mean_p - mean.t0$mean_p

ipw.reg <- lm(price ~ penalty, data=pen.data.2012, weights=ipw)


## Regression
reg.data <- pen.data.2012 %>% ungroup() %>%
  mutate(size1_diff = penalty*(bed_size1 - mean(bed_size1)),
         size2_diff = penalty*(bed_size2 - mean(bed_size2)),
         size3_diff = penalty*(bed_size3 - mean(bed_size3)))

reg <- lm(price ~ penalty + bed_size1 + bed_size2 + bed_size3 +
            size1_diff + size2_diff + size3_diff,
          data=reg.data)


rm(list=c("version.1996", "version.2010", "pen.data", "pen.data.2012",
          "hcris.data.v1996", "hcris.data.v2010", "charge.data", "price.data",
          "final.hcris.v1996", "final.hcris", "hcris.data", "reg.data"))
save.image("assignments/Hwk2_workspace.Rdata")

File: analysis/hwk2-analysis.py

# Meta --------------------------------------------------------------------

## Title:         Econ/HLTH 470 Homework 2 Answers
## Author:        Pablo Estrada
## Date Created:  2/10/2025
## Date Edited:   2/10/2025
## Description:   This file renders/runs Python code for the assignment


# Preliminaries -----------------------------------------------------------

# Importing the libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from causalinference import CausalModel
blue = '#002878'
gold = '#d2b000'

# Read data
hcris_data_v1996 = pd.read_csv('data/output/HCRIS_v1996.csv')
hcris_data_v2010 = pd.read_csv('data/output/HCRIS_v2010.csv')
hcris_data = pd.read_csv('data/output/HCRIS_Data.csv')

# Filter data for the years 2008-2015
hcris_data = hcris_data[(hcris_data['year'] >= 2008) & (hcris_data['year'] <= 2015)]

# Create summary objects
version_1996 = hcris_data_v1996.groupby('year').size().reset_index(name='count_1996')
version_2010 = hcris_data_v2010.groupby('year').size().reset_index(name='count_2010')

# Merge datasets
version_data = pd.merge(version_1996, version_2010, on='year', how='outer')

# Add missing columns in v1996 dataset
hcris_data_v1996['hvbp_payment'] = np.nan
hcris_data_v1996['hrrp_payment'] = np.nan

# Combine v1996 and v2010 datasets
final_hcris = pd.concat([hcris_data_v1996, hcris_data_v2010], ignore_index=True)

# Convert date columns to datetime format
date_cols = ['fy_end', 'fy_start', 'date_processed', 'date_created']
for col in date_cols:
    final_hcris[col] = pd.to_datetime(final_hcris[col], errors='coerce')

# Apply transformations
final_hcris['tot_discounts'] = final_hcris['tot_discounts'].abs()
final_hcris['hrrp_payment'] = final_hcris['hrrp_payment'].abs()
final_hcris['fyear'] = final_hcris['fy_end'].dt.year

# Sort and filter
final_hcris = final_hcris.sort_values(by=['provider_number', 'fyear'])
final_hcris = final_hcris[(final_hcris['fyear'] >= 2008) & (final_hcris['fyear'] <= 2015)]

# Count distinct providers
num_providers = hcris_data['provider_number'].nunique()

# Count duplicate reports per provider per year
dup_count = final_hcris.groupby(['provider_number', 'fyear']).size().reset_index(name='total_reports')
dup_count['dup_report'] = dup_count['total_reports'] > 1

# Summarize duplicates per year
dup_summary = dup_count.groupby('fyear')['dup_report'].sum().reset_index()

# Plot duplicate reports per year
plt.figure(figsize=(8, 5))
plt.plot(dup_summary['fyear'].values, dup_summary['dup_report'].values, 
         marker='o', linestyle='-', color=blue)

# Format plot
plt.title('Figure 1: Duplicate Reports', fontsize=14)
plt.xlabel('')
plt.ylabel('Number of Hospitals')
plt.ylim(0, 300)
plt.grid(axis='y', color='gray', linestyle='--', alpha=0.5)
plt.show()

# Count unique hospitals per year
hosp_count = hcris_data.groupby('year').size().reset_index(name='hosp_count')

# Plot unique hospitals per year
plt.figure(figsize=(8, 5))
plt.plot(hosp_count['year'].values, hosp_count['hosp_count'].values, 
         marker='o', linestyle='-', color=blue)

# Format plot
plt.title('Figure 2: Unique Hospitals', fontsize=14)
plt.xlabel('')
plt.ylabel('Number of Hospitals')
plt.ylim(0, 8000)
plt.grid(axis='y', color='gray', linestyle='--', alpha=0.5)
plt.show()

# Filter and preprocess data
charge_data = hcris_data.copy()

# Compute 1st and 99th percentiles for total charges per year
charge_data['tot_charges_low'] = charge_data.groupby('year')['tot_charges'].transform(lambda x: np.nanpercentile(x, 1))
charge_data['tot_charges_high'] = charge_data.groupby('year')['tot_charges'].transform(lambda x: np.nanpercentile(x, 99))

# Filter out extreme values and missing data
charge_data = charge_data[
    (charge_data['tot_charges'] > charge_data['tot_charges_low']) &
    (charge_data['tot_charges'] < charge_data['tot_charges_high']) &
    charge_data['tot_charges'].notna() &
    (charge_data['year'] > 1997)
]

# Compute log of total charges
charge_data['log_charge'] = np.log(charge_data['tot_charges'])

# Preparing data for violin plot
years = sorted(charge_data['year'].unique())
data = [charge_data[charge_data['year'] == y]['log_charge'].dropna().values for y in years]

# Plot distribution of total charges
fig, ax = plt.subplots(figsize=(8, 5))
parts = ax.violinplot(data, positions=years, showmedians=False)

# Customize violin plot appearance
for pc in parts['bodies']:
    pc.set_facecolor(gold)
    pc.set_alpha(0.3)

for part in ['cbars', 'cmins', 'cmaxes']:
    parts[part].set_color(blue)
    parts[part].set_linewidth(0.5)

# Add median line
median = charge_data.groupby('year')['log_charge'].median()
plt.plot(years, median.values, marker='o', linestyle='-', color=blue, linewidth=2)

# Format plot
plt.title('Distribution of Total Charges', fontsize=14)
plt.xlabel('')
plt.ylabel('Log \$')
plt.grid(axis='y', color='gray', linestyle='--', alpha=0.5)
plt.show()

# Compute price data
hcris_data['discount_factor'] = 1 - hcris_data['tot_discounts'] / hcris_data['tot_charges']
hcris_data['price_num'] = (
    (hcris_data['ip_charges'] + hcris_data['icu_charges'] + hcris_data['ancillary_charges']) * 
    hcris_data['discount_factor']
) - hcris_data['tot_mcare_payment']
hcris_data['price_denom'] = hcris_data['tot_discharges'] - hcris_data['mcare_discharges']
hcris_data['price'] = hcris_data['price_num'] / hcris_data['price_denom']

# Filtering the data
price_data = hcris_data[
    (hcris_data['price_denom'] > 100) & 
    (~hcris_data['price_denom'].isna()) & 
    (hcris_data['price_num'] > 0) & 
    (~hcris_data['price_num'].isna()) & 
    (hcris_data['price'] < 100000) & 
    (hcris_data['beds'] > 30) & 
    (~hcris_data['beds'].isna())
]

# Preparing data for violin plot
years = sorted(price_data['year'].unique())
data = [price_data[price_data['year'] == year]['price'].dropna().values for year in years]

# Plot distribution of hospital prices
fig, ax = plt.subplots(figsize=(8, 5))
parts = ax.violinplot(data, positions=years, showmedians=False, showextrema=False)

# Customize violin plot appearance
for pc in parts['bodies']:
    pc.set_facecolor(gold)
    pc.set_alpha(0.3)

# Add median line
median = price_data.groupby('year')['price'].median()
plt.plot(years, median.values, marker='o', linestyle='-', color=blue, linewidth=2)

# Format plot
plt.title('Distribution of Hospital Prices', fontsize=14)
plt.xlabel('')
plt.ylabel('Price')
plt.grid(axis='y', color='gray', linestyle='--', alpha=0.5)
plt.show()

# Compute penalty data
price_data.loc[:, 'penalty'] = ((price_data['hvbp_payment'].fillna(0) - price_data['hrrp_payment'].fillna(0).abs()) < 0).astype(int)

# Filter data for the year 2012
pen_data_2012 = price_data[price_data['year'] == 2012].copy()

# Compute quartiles
beds_q1 = pen_data_2012['beds'].quantile(0.25)
beds_q2 = pen_data_2012['beds'].quantile(0.50)
beds_q3 = pen_data_2012['beds'].quantile(0.75)
beds_q4 = pen_data_2012['beds'].max()

# Assign bed quartiles
pen_data_2012['bed_quart'] = np.select(
    [
        pen_data_2012['beds'] < beds_q1,
        (pen_data_2012['beds'] >= beds_q1) & (pen_data_2012['beds'] < beds_q2),
        (pen_data_2012['beds'] >= beds_q2) & (pen_data_2012['beds'] < beds_q3),
        (pen_data_2012['beds'] > beds_q3) & (pen_data_2012['beds'] < beds_q4)
    ],
    [1, 2, 3, 4],
    default=0
)

# Filter out invalid quartile values
pen_data_2012 = pen_data_2012[pen_data_2012['bed_quart'] > 0]

# Compute average price by penalty status
avg_pen = pen_data_2012.groupby('penalty')['price'].mean()

# Compute average price by penalty status and bed quartile
avg_pen = pen_data_2012.groupby(['penalty', 'bed_quart'])[['price']].mean().unstack(level=0)

# Change the name of the index and columns
avg_pen.columns = avg_pen.columns.droplevel()
avg_pen.columns = ['Not Penalized', 'Penalized']
avg_pen.columns.name = ''
avg_pen.index.name = 'Bed Size'

# Create bed quartile dummies for models
bed_quarts = pd.get_dummies(pen_data_2012['bed_quart'], prefix='bed_quart').iloc[:, :-1] * 1
bed_quarts = bed_quarts.sub(bed_quarts.mean(axis=0), axis=1)

cm = CausalModel(
    Y = pen_data_2012['price'].values,
    D = pen_data_2012['penalty'].values,
    # X = pd.get_dummies(pen_data_2012['bed_quart']).values * 1
    X = bed_quarts.values
)

# print(cm.summary_stats)
results = pd.DataFrame(index=['ATE', 'SE'], columns=['INV', 'MAH', 'IPW', 'OLS'])

# Matching by inverse variance
cm.est_via_matching(weights='inv', matches=1, bias_adj=True)
results.loc['ATE', 'INV'] = cm.estimates['matching']['ate']
results.loc['SE', 'INV'] = cm.estimates['matching']['ate_se']

# Matching by Mahalanobis distance
cm.est_via_matching(weights='maha', matches=1, bias_adj=True)
results.loc['ATE', 'MAH'] = cm.estimates['matching']['ate']
results.loc['SE', 'MAH'] = cm.estimates['matching']['ate_se']

# Inverse probability weighting
cm.est_propensity()
cm.est_via_weighting()
results.loc['ATE', 'IPW'] = cm.estimates['weighting']['ate']
results.loc['SE', 'IPW'] = cm.estimates['weighting']['ate_se']

# OLS Regression
cm.est_via_ols(adj=2)  # 1 for linear model, 2 for interaction model
results.loc['ATE', 'OLS'] = cm.estimates['ols']['ate']
results.loc['SE', 'OLS'] = cm.estimates['ols']['ate_se']
